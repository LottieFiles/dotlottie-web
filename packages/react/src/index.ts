export * from './dotlottie';
export type * from '@lottiefiles/dotlottie-web';
export * from './use-dotlottie';
