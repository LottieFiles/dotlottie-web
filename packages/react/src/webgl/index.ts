export * from './dotlottie';
export type * from '@lottiefiles/dotlottie-web/webgl';
