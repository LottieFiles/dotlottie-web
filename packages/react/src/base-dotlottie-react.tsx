/* eslint-disable no-warning-comments */
'use client';

import type { Config, DotLottie, DotLottieWorker } from '@lottiefiles/dotlottie-web';
import { useState, useEffect, useCallback, useRef, type ComponentProps, type RefCallback } from 'react';
import type { JSX } from 'react';

export type BaseDotLottieProps<T extends DotLottie | DotLottieWorker> = Omit<Config, 'canvas'> &
  ComponentProps<'canvas'> & {
    animationId?: string;
    /**
     * A function that creates a `DotLottie` or `DotLottieWorker` instance.
     */
    createDotLottie: (config: T extends DotLottieWorker ? Config & { workerId?: string } : Config) => T;
    /**
     * A callback function that receives the `DotLottie` or `DotLottieWorker` instance.
     *
     * @example
     * ```tsx
     * const [dotLottie, setDotLottie] = useState<DotLottie | null>(null);
     *
     * <DotLottieReact
     *   dotLottieRefCallback={setDotLottie}
     * />
     * ```
     */
    dotLottieRefCallback?: RefCallback<T | null>;
    /**
     * @deprecated The `playOnHover` property is deprecated.
     * Instead, use the `onMouseEnter` and `onMouseLeave` events to control animation playback.
     * Utilize the `dotLottieRefCallback` to access the `DotLottie` instance and invoke the `play` and `pause` methods.
     *
     * Example usage:
     * ```tsx
     * const [dotLottie, setDotLottie] = useState<DotLottie | null>(null);
     *
     * <DotLottieReact
     *   dotLottieRefCallback={setDotLottie}
     *   onMouseEnter={() => dotLottie?.play()}
     *   onMouseLeave={() => dotLottie?.pause()}
     * />
     * ```
     */
    playOnHover?: boolean;
    themeData?: string;
    workerId?: T extends DotLottieWorker ? string : undefined;
  };

export const BaseDotLottieReact = <T extends DotLottie | DotLottieWorker>({
  animationId,
  autoplay,
  backgroundColor,
  className,
  createDotLottie,
  data,
  dotLottieRefCallback,
  layout,
  loop,
  mode,
  playOnHover,
  renderConfig,
  segment,
  speed,
  src,
  style,
  themeData,
  themeId,
  useFrameInterpolation,
  workerId,
  ...props
}: BaseDotLottieProps<T> & {
  createDotLottie: (config: T extends DotLottieWorker ? Config & { workerId?: string } : Config) => T;
}): JSX.Element => {
  const [dotLottie, setDotLottie] = useState<T | null>(null);
  const canvasRef = useRef<HTMLCanvasElement | null>(null);
  const dotLottieRef = useRef<T | null>(null);
  const dotLottieRefCallbackRef = useRef<RefCallback<T | null> | undefined>(dotLottieRefCallback);

  const config: Omit<Config, 'canvas'> & {
    workerId?: T extends DotLottieWorker ? string : undefined;
  } = {
    speed,
    mode,
    loop,
    useFrameInterpolation,
    segment,
    backgroundColor,
    autoplay,
    themeId,
    workerId,
    src,
    data,
    layout,
    renderConfig,
  };

  const configRef = useRef<Omit<BaseDotLottieProps<T>, 'createDotLottie' | 'dotLottieRefCallback'> | undefined>(config);

  dotLottieRefCallbackRef.current = dotLottieRefCallback;
  dotLottieRef.current = dotLottie;
  configRef.current = config;

  useEffect(() => {
    if (typeof dotLottieRefCallbackRef.current === 'function' && dotLottie) {
      dotLottieRefCallbackRef.current(dotLottie);
    }
  }, [dotLottie]);

  const setCanvasRef = useCallback((canvas: HTMLCanvasElement | null) => {
    canvasRef.current = canvas;

    if (canvas) {
      const dotLottieInstance = createDotLottie({
        ...configRef.current,
        canvas,
      });

      setDotLottie(dotLottieInstance as T);
    } else {
      dotLottie?.destroy();
      setDotLottie(null);
    }
  }, []);

  useEffect(() => {
    const handlePlayOnHover = (event: MouseEvent): void => {
      if (!playOnHover) return;

      if (event.type === 'mouseenter') {
        dotLottieRef.current?.play();
      }

      if (event.type === 'mouseleave') {
        dotLottieRef.current?.pause();
      }
    };

    canvasRef.current?.addEventListener('mouseenter', handlePlayOnHover);
    canvasRef.current?.addEventListener('mouseleave', handlePlayOnHover);

    return () => {
      canvasRef.current?.removeEventListener('mouseenter', handlePlayOnHover);
      canvasRef.current?.removeEventListener('mouseleave', handlePlayOnHover);
    };
  }, [playOnHover]);

  useEffect(() => {
    return () => {
      if (dotLottie) {
        dotLottie.destroy();
        setDotLottie(null);
      }
    };
  }, [dotLottie]);

  useEffect(() => {
    dotLottieRef.current?.setSpeed(speed ?? 1);
  }, [speed]);

  useEffect(() => {
    dotLottieRef.current?.setMode(mode ?? 'forward');
  }, [mode]);

  useEffect(() => {
    dotLottieRef.current?.setLoop(loop ?? false);
  }, [loop]);

  useEffect(() => {
    dotLottieRef.current?.setUseFrameInterpolation(useFrameInterpolation ?? true);
  }, [useFrameInterpolation]);

  useEffect(() => {
    if (typeof segment?.[0] === 'number' && typeof segment[1] === 'number') {
      dotLottieRef.current?.setSegment(segment[0], segment[1]);
    } else {
      // TODO: implement it for worker
      // dotLottieRef.current?.resetSegment();
    }
  }, [segment]);

  useEffect(() => {
    dotLottieRef.current?.setBackgroundColor(backgroundColor ?? '');
  }, [backgroundColor]);

  useEffect(() => {
    dotLottieRef.current?.setRenderConfig(renderConfig ?? {});
  }, [JSON.stringify(renderConfig)]);

  useEffect(() => {
    if (typeof data !== 'string' && typeof data !== 'object') return;

    dotLottieRef.current?.load({
      data,
      ...configRef.current,
    });
  }, [data]);

  useEffect(() => {
    if (typeof src !== 'string') return;

    dotLottieRef.current?.load({
      src,
      ...configRef.current,
    });
  }, [src]);

  useEffect(() => {
    dotLottieRef.current?.setMarker(props.marker ?? '');
  }, [props.marker]);

  useEffect(() => {
    dotLottieRef.current?.loadAnimation(animationId ?? '');
  }, [animationId]);

  useEffect(() => {
    if (typeof themeId === 'string') {
      dotLottieRef.current?.setTheme(themeId);
    } else {
      // TODO: implement it for worker
      // dotLottieRef.current?.resetTheme();
    }
  }, [themeId]);

  useEffect(() => {
    dotLottieRef.current?.setThemeData(themeData ?? '');
  }, [themeData]);

  useEffect(() => {
    dotLottieRef.current?.setLayout({
      align: layout?.align ?? [0.5, 0.5],
      fit: layout?.fit ?? 'contain',
    });
  }, [layout?.align[0], layout?.align[1], layout?.fit]);

  return (
    <div
      className={className}
      {...(!className && {
        style: {
          width: '100%',
          height: '100%',
          lineHeight: 0,
          ...style,
        },
      })}
    >
      <canvas
        ref={setCanvasRef}
        style={{
          width: '100%',
          height: '100%',
        }}
        {...props}
      />
    </div>
  );
};
