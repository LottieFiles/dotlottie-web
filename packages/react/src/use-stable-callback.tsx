/**
 * Copyright 2024 Design Barn Inc.
 */

import { useCallback, useLayoutEffect, useRef, useEffect } from 'react';

// eslint-disable-next-line no-negated-condition
const useIsomorphicLayoutEffect = typeof window !== 'undefined' ? useLayoutEffect : useEffect;

/**
 * Returns a memoized version of your function that maintains a stable reference, but
 * also can read the latest scope (props and state) of the component in which it is used.
 */
export default function useStableCallback<Args extends unknown[], T>(
  callback: (...args: Args) => T,
): (...args: Args) => T {
  const callbackContainer = useRef(callback);

  useIsomorphicLayoutEffect(() => {
    callbackContainer.current = callback;
  });

  return useCallback((...args: Args) => callbackContainer.current(...args), [callbackContainer]);
}
