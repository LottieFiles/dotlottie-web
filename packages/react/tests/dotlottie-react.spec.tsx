/* eslint-disable node/no-unsupported-features/node-builtins */
import { DotLottie, DotLottieWorker } from '@lottiefiles/dotlottie-web';
import { DotLottie as DotLottieWebGL } from '@lottiefiles/dotlottie-web/webgl';
import { DotLottie as DotLottieWebGPU } from '@lottiefiles/dotlottie-web/webgpu';
import { userEvent } from '@testing-library/user-event';
import React from 'react';
import { afterEach, describe, expect, test, vi } from 'vitest';
import type { ComponentRenderOptions, RenderResult } from 'vitest-browser-react';
import { cleanup, render as vitestRender } from 'vitest-browser-react';

import { DotLottieReact, DotLottieWorkerReact, setWasmUrl } from '../src';
import { DotLottieReact as DotLottieWebGLReact, setWasmUrl as setWebGLWasmUrl } from '../src/webgl';
import { DotLottieReact as DotLottieWebGPUReact, setWasmUrl as setWebGPUWasmUrl } from '../src/webgpu';

setWasmUrl(new URL('../../web/src/software/wasm/dotlottie-player.wasm?url', import.meta.url).href);
setWebGLWasmUrl(new URL('../../web/src/webgl/wasm/dotlottie-player.wasm?url', import.meta.url).href);
setWebGPUWasmUrl(new URL('../../web/src/webgpu/wasm/dotlottie-player.wasm?url', import.meta.url).href);

const dotLottieSrc = new URL('./__fixtures__/test.lottie', import.meta.url).href;
const lottieSrc = new URL('./__fixtures__/test.json', import.meta.url).href;

const Wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  return <>{children}</>;
};

const render = (ui: React.ReactNode, options?: ComponentRenderOptions): RenderResult =>
  vitestRender(ui, { wrapper: Wrapper, ...options });

describe.each([
  { name: 'DotLottieReact', component: DotLottieReact, instanceType: DotLottie },
  { name: 'DotLottieWorkerReact', component: DotLottieWorkerReact, instanceType: DotLottieWorker },
  { name: 'DotLottieWebGPUReact', component: DotLottieWebGPUReact, instanceType: DotLottieWebGPU },
  { name: 'DotLottieWebGLReact', component: DotLottieWebGLReact, instanceType: DotLottieWebGL },
])('$name', ({ component: Component, instanceType }) => {
  afterEach(() => {
    cleanup();
  });

  test('basic', async () => {
    const onLoad = vi.fn();
    const onDestroy = vi.fn();
    const onComplete = vi.fn();
    const dotLottieRefCallback = vi.fn();

    const { container, unmount } = render(
      <Component src={dotLottieSrc} autoplay dotLottieRefCallback={dotLottieRefCallback} />,
    );

    expect(container).toMatchSnapshot();

    await vi.waitFor(() => {
      expect(dotLottieRefCallback).toHaveBeenCalledTimes(1);
    });

    const dotLottie = dotLottieRefCallback.mock.calls[0]?.[0];

    expect(dotLottie).toBeInstanceOf(instanceType);

    dotLottie?.addEventListener('load', onLoad);
    dotLottie?.addEventListener('destroy', onDestroy);
    dotLottie?.addEventListener('complete', onComplete);

    await vi.waitFor(() => {
      expect(onLoad).toHaveBeenCalledTimes(1);
    });

    await vi.waitFor(
      () => {
        expect(onComplete).toHaveBeenCalledTimes(1);
      },
      { timeout: dotLottie?.duration * 1000 + 100 },
    );

    unmount();

    await vi.waitFor(() => {
      expect(onDestroy).toHaveBeenCalledTimes(1);
    });
  });

  test('calls dotLottie.setLoop when loop prop changes', async () => {
    const onLoad = vi.fn();
    const dotLottieRefCallback = vi.fn();

    const { rerender } = render(<Component src={dotLottieSrc} autoplay dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(dotLottieRefCallback).toHaveBeenCalledTimes(1);
    });

    const dotLottie = dotLottieRefCallback.mock.calls[0]?.[0];

    const setLoop = vi.spyOn(dotLottie, 'setLoop');

    dotLottie?.addEventListener('load', onLoad);

    await vi.waitFor(() => {
      expect(onLoad).toHaveBeenCalledTimes(1);
    });

    expect(dotLottie?.loop).toBe(false);

    rerender(<Component src={dotLottieSrc} autoplay loop dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(setLoop).toHaveBeenCalledTimes(1);
    });

    expect(setLoop).toHaveBeenCalledWith(true);

    await vi.waitFor(() => {
      expect(dotLottie?.loop).toBe(true);
    });

    rerender(<Component src={dotLottieSrc} autoplay dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(setLoop).toHaveBeenCalledTimes(2);
    });

    expect(setLoop).toHaveBeenCalledWith(false);

    await vi.waitFor(() => {
      expect(dotLottie?.loop).toBe(false);
    });
  });

  test('calls dotLottie.setSpeed when speed prop changes', async () => {
    const onLoad = vi.fn();
    const dotLottieRefCallback = vi.fn();

    const { rerender } = render(<Component src={dotLottieSrc} autoplay dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(dotLottieRefCallback).toHaveBeenCalledTimes(1);
    });

    const dotLottie = dotLottieRefCallback.mock.calls[0]?.[0];

    dotLottie?.addEventListener('load', onLoad);

    await vi.waitFor(() => {
      expect(onLoad).toHaveBeenCalledTimes(1);
    });

    const setSpeed = vi.spyOn(dotLottie, 'setSpeed');

    expect(dotLottie?.speed).toBe(1);

    rerender(<Component src={dotLottieSrc} autoplay speed={2} dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(setSpeed).toHaveBeenCalledTimes(1);
    });

    expect(setSpeed).toHaveBeenCalledWith(2);

    await vi.waitFor(() => {
      expect(dotLottie?.speed).toBe(2);
    });

    rerender(<Component src={dotLottieSrc} autoplay dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(setSpeed).toHaveBeenCalledTimes(2);
    });

    expect(setSpeed).toHaveBeenCalledWith(1);

    await vi.waitFor(() => {
      expect(dotLottie?.speed).toBe(1);
    });
  });

  test('calls dotLottie.setMode when mode prop changes', async () => {
    const onLoad = vi.fn();
    const dotLottieRefCallback = vi.fn();

    const { rerender } = render(<Component src={dotLottieSrc} autoplay dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(dotLottieRefCallback).toHaveBeenCalledTimes(1);
    });

    const dotLottie = dotLottieRefCallback.mock.calls[0]?.[0];

    dotLottie?.addEventListener('load', onLoad);

    await vi.waitFor(() => {
      expect(onLoad).toHaveBeenCalledTimes(1);
    });

    const setMode = vi.spyOn(dotLottie, 'setMode');

    expect(dotLottie?.mode).toBe('forward');

    rerender(<Component src={dotLottieSrc} autoplay mode="reverse" dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(setMode).toHaveBeenCalledTimes(1);
    });

    expect(setMode).toHaveBeenCalledWith('reverse');

    await vi.waitFor(() => {
      expect(dotLottie?.mode).toBe('reverse');
    });

    rerender(<Component src={dotLottieSrc} autoplay dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(setMode).toHaveBeenCalledTimes(2);
    });

    expect(setMode).toHaveBeenCalledWith('forward');

    await vi.waitFor(() => {
      expect(dotLottie?.mode).toBe('forward');
    });
  });

  test('calls dotLottie.setUseFrameInterpolation when useFrameInterpolation prop changes', async () => {
    const onLoad = vi.fn();
    const dotLottieRefCallback = vi.fn();

    const { rerender } = render(<Component src={dotLottieSrc} autoplay dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(dotLottieRefCallback).toHaveBeenCalledTimes(1);
    });

    const dotLottie = dotLottieRefCallback.mock.calls[0]?.[0];

    const setUseFrameInterpolation = vi.spyOn(dotLottie, 'setUseFrameInterpolation');

    dotLottie?.addEventListener('load', onLoad);

    await vi.waitFor(() => {
      expect(onLoad).toHaveBeenCalledTimes(1);
    });

    expect(dotLottie?.useFrameInterpolation).toBe(true);

    rerender(
      <Component
        src={dotLottieSrc}
        autoplay
        useFrameInterpolation={false}
        dotLottieRefCallback={dotLottieRefCallback}
      />,
    );

    await vi.waitFor(() => {
      expect(setUseFrameInterpolation).toHaveBeenCalledTimes(1);
    });

    expect(setUseFrameInterpolation).toHaveBeenCalledWith(false);

    await vi.waitFor(() => {
      expect(dotLottie?.useFrameInterpolation).toBe(false);
    });

    rerender(<Component src={dotLottieSrc} autoplay dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(setUseFrameInterpolation).toHaveBeenCalledTimes(2);
    });

    expect(setUseFrameInterpolation).toHaveBeenCalledWith(true);

    await vi.waitFor(() => {
      expect(dotLottie?.useFrameInterpolation).toBe(true);
    });
  });

  test('calls dotLottie.setBackgroundColor when backgroundColor prop changes', async () => {
    const onLoad = vi.fn();
    const dotLottieRefCallback = vi.fn();

    const { rerender } = render(<Component src={dotLottieSrc} autoplay dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(dotLottieRefCallback).toHaveBeenCalledTimes(1);
    });

    const dotLottie = dotLottieRefCallback.mock.calls[0]?.[0];

    const setBackgroundColor = vi.spyOn(dotLottie, 'setBackgroundColor');

    dotLottie?.addEventListener('load', onLoad);

    await vi.waitFor(() => {
      expect(onLoad).toHaveBeenCalledTimes(1);
    });

    expect(dotLottie?.backgroundColor).toBe('');

    rerender(
      <Component src={dotLottieSrc} autoplay backgroundColor="#00ff00ff" dotLottieRefCallback={dotLottieRefCallback} />,
    );

    await vi.waitFor(() => {
      expect(setBackgroundColor).toHaveBeenCalledTimes(1);
    });

    expect(setBackgroundColor).toHaveBeenCalledWith('#00ff00ff');

    rerender(<Component src={dotLottieSrc} autoplay dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(setBackgroundColor).toHaveBeenCalledTimes(2);
    });

    expect(dotLottie?.backgroundColor).toBe('');
  });

  test('calls dotLottie.setMarker when marker prop changes', async () => {
    const onLoad = vi.fn();
    const dotLottieRefCallback = vi.fn();

    const { rerender } = render(<Component src={dotLottieSrc} autoplay dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(dotLottieRefCallback).toHaveBeenCalledTimes(1);
    });

    const dotLottie = dotLottieRefCallback.mock.calls[0]?.[0];

    const setMarker = vi.spyOn(dotLottie, 'setMarker');

    dotLottie?.addEventListener('load', onLoad);

    await vi.waitFor(() => {
      expect(onLoad).toHaveBeenCalledTimes(1);
    });

    expect(dotLottie?.marker).toBe('');

    rerender(<Component src={dotLottieSrc} autoplay marker="Marker_1" dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(setMarker).toHaveBeenCalledTimes(1);
    });

    expect(setMarker).toHaveBeenCalledWith('Marker_1');

    rerender(<Component src={dotLottieSrc} autoplay dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(setMarker).toHaveBeenCalledTimes(2);
    });

    expect(dotLottie?.marker).toBe('');
  });

  test.todo('calls dotLottie.setSegment & dotLottie.resetSegment when segment prop changes', async () => {
    const onLoad = vi.fn();
    const dotLottieRefCallback = vi.fn();

    const { rerender } = render(<Component src={dotLottieSrc} autoplay dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(dotLottieRefCallback).toHaveBeenCalledTimes(1);
    });

    const dotLottie = dotLottieRefCallback.mock.calls[0]?.[0];

    const setSegment = vi.spyOn(dotLottie, 'setSegment');
    const resetSegment = vi.spyOn(dotLottie, 'resetSegment');

    dotLottie?.addEventListener('load', onLoad);

    await vi.waitFor(() => {
      expect(onLoad).toHaveBeenCalledTimes(1);
    });

    expect(dotLottie?.segment).toBeUndefined();

    rerender(<Component src={dotLottieSrc} autoplay segment={[0, 10]} dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(setSegment).toHaveBeenCalledTimes(1);
    });

    expect(setSegment).toHaveBeenCalledWith(0, 10);

    rerender(<Component src={dotLottieSrc} autoplay dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(resetSegment).toHaveBeenCalledTimes(1);
    });

    expect(dotLottie?.segment).toBeUndefined();
  });

  test.todo('calls dotLottie.setTheme & dotLottie.resetTheme when themeId prop changes', async () => {
    const onLoad = vi.fn();
    const dotLottieRefCallback = vi.fn();

    const { rerender } = render(<Component src={dotLottieSrc} autoplay dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(dotLottieRefCallback).toHaveBeenCalledTimes(1);
    });

    const dotLottie = dotLottieRefCallback.mock.calls[0]?.[0];

    const setTheme = vi.spyOn(dotLottie, 'setTheme');
    const resetTheme = vi.spyOn(dotLottie, 'resetTheme');

    dotLottie?.addEventListener('load', onLoad);

    await vi.waitFor(() => {
      expect(onLoad).toHaveBeenCalledTimes(1);
    });

    expect(dotLottie?.themeId).toBeUndefined();

    rerender(<Component src={dotLottieSrc} autoplay themeId="Theme_1" dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(setTheme).toHaveBeenCalledTimes(1);
    });

    expect(setTheme).toHaveBeenCalledWith('Theme_1');

    rerender(<Component src={dotLottieSrc} autoplay dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(resetTheme).toHaveBeenCalledTimes(1);
    });

    expect(dotLottie?.themeId).toBeUndefined();
  });

  test('playOnHover', async () => {
    const user = userEvent.setup();

    const onLoad = vi.fn();
    const dotLottieRefCallback = vi.fn();

    const screen = render(
      <Component
        data-testid="dotLottie-canvas"
        src={dotLottieSrc}
        autoplay
        playOnHover
        dotLottieRefCallback={dotLottieRefCallback}
      />,
    );

    const dotLottie = dotLottieRefCallback.mock.calls[0]?.[0];

    dotLottie?.addEventListener('load', onLoad);

    await vi.waitFor(() => {
      expect(onLoad).toHaveBeenCalledTimes(1);
    });

    const play = vi.spyOn(dotLottie, 'play');
    const pause = vi.spyOn(dotLottie, 'pause');

    let canvasElement = screen.getByTestId('dotLottie-canvas').element();

    await user.hover(canvasElement);

    await vi.waitFor(() => {
      expect(play).toHaveBeenCalledTimes(1);
    });

    await user.unhover(canvasElement);

    await vi.waitFor(() => {
      expect(pause).toHaveBeenCalledTimes(1);
    });

    play.mockClear();
    pause.mockClear();

    // shouldn't call play/pause again as the playOnHover prop is undefined
    screen.rerender(
      <Component
        data-testid="dotLottie-canvas"
        src={dotLottieSrc}
        autoplay
        dotLottieRefCallback={dotLottieRefCallback}
      />,
    );

    canvasElement = screen.getByTestId('dotLottie-canvas').element();

    let mouseEnterCount = 0;
    let mouseLeaveCount = 0;

    canvasElement.addEventListener('mouseenter', () => {
      mouseEnterCount += 1;
    });
    canvasElement.addEventListener('mouseleave', () => {
      mouseLeaveCount += 1;
    });

    await user.hover(canvasElement);

    await vi.waitFor(() => {
      expect(mouseEnterCount).toBe(1);
      expect(mouseLeaveCount).toBe(0);
    });

    expect(play).not.toHaveBeenCalled();

    await user.unhover(canvasElement);

    await vi.waitFor(() => {
      expect(mouseEnterCount).toBe(1);
      expect(mouseLeaveCount).toBe(1);
    });

    expect(pause).not.toHaveBeenCalled();
  });

  test('calls dotLottie.loadAnimation when animationId prop changes', async () => {
    const onLoad = vi.fn();
    const dotLottieRefCallback = vi.fn();

    const { rerender } = render(<Component src={dotLottieSrc} autoplay dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(dotLottieRefCallback).toHaveBeenCalledTimes(1);
    });

    const dotLottie = dotLottieRefCallback.mock.calls[0]?.[0];

    dotLottie?.addEventListener('load', onLoad);

    await vi.waitFor(() => {
      expect(onLoad).toHaveBeenCalledTimes(1);
    });

    const loadAnimation = vi.spyOn(dotLottie, 'loadAnimation');

    rerender(
      <Component src={dotLottieSrc} autoplay animationId="Animation_1" dotLottieRefCallback={dotLottieRefCallback} />,
    );

    await vi.waitFor(() => {
      expect(loadAnimation).toHaveBeenCalledTimes(1);
    });

    expect(loadAnimation).toHaveBeenCalledWith('Animation_1');

    rerender(<Component src={dotLottieSrc} autoplay dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(loadAnimation).toHaveBeenCalledTimes(2);
    });

    expect(loadAnimation).toHaveBeenCalledWith('');

    await vi.waitFor(() => {
      expect(dotLottie?.activeAnimationId).toBe('');
    });
  });

  test('calls dotLottie.setRenderConfig when renderConfig prop changes', async () => {
    const onLoad = vi.fn();
    const dotLottieRefCallback = vi.fn();

    const { rerender } = render(<Component src={dotLottieSrc} autoplay dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(dotLottieRefCallback).toHaveBeenCalledTimes(1);
    });

    const dotLottie = dotLottieRefCallback.mock.calls[0]?.[0];

    dotLottie?.addEventListener('load', onLoad);

    await vi.waitFor(() => {
      expect(onLoad).toHaveBeenCalledTimes(1);
    });

    const defaultRenderConfig = dotLottie?.renderConfig;

    const setRenderConfig = vi.spyOn(dotLottie, 'setRenderConfig');

    rerender(
      <Component
        src={dotLottieSrc}
        autoplay
        renderConfig={{
          devicePixelRatio: 0.5,
          freezeOnOffscreen: false,
        }}
        dotLottieRefCallback={dotLottieRefCallback}
      />,
    );

    await vi.waitFor(() => {
      expect(setRenderConfig).toHaveBeenCalledTimes(1);
    });

    expect(setRenderConfig).toHaveBeenCalledWith({
      devicePixelRatio: 0.5,
      freezeOnOffscreen: false,
    });

    rerender(<Component src={dotLottieSrc} autoplay dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(setRenderConfig).toHaveBeenCalledTimes(2);
    });

    expect(setRenderConfig).toHaveBeenCalledWith({});

    // Falls back to the default values
    expect(dotLottie?.renderConfig).toEqual(defaultRenderConfig);
  });

  test('calls dotLottie.load when data prop changes', async () => {
    const onLoad = vi.fn();
    const dotLottieRefCallback = vi.fn();

    let response = await fetch(lottieSrc);
    const animationData = await response.json();

    const { rerender } = render(
      <Component data={animationData} autoplay dotLottieRefCallback={dotLottieRefCallback} />,
    );

    await vi.waitFor(() => {
      expect(dotLottieRefCallback).toHaveBeenCalledTimes(1);
    });

    const dotLottie = dotLottieRefCallback.mock.calls[0]?.[0];

    dotLottie?.addEventListener('load', onLoad);

    await vi.waitFor(() => {
      expect(onLoad).toHaveBeenCalledTimes(1);
    });

    const load = vi.spyOn(dotLottie, 'load');

    response = await fetch(dotLottieSrc);
    const dotLottieAnimationData = await response.arrayBuffer();

    rerender(
      <Component data={dotLottieAnimationData} autoplay loop speed={2} dotLottieRefCallback={dotLottieRefCallback} />,
    );

    await vi.waitFor(() => {
      expect(load).toHaveBeenCalledTimes(1);
    });

    expect(load).toHaveBeenCalledWith({
      data: dotLottieAnimationData,
      loop: true,
      autoplay: true,
      speed: 2,
    });
  });

  test('calls dotLottie.load when src prop changes', async () => {
    const onLoad = vi.fn();
    const dotLottieRefCallback = vi.fn();

    const { rerender } = render(<Component src={dotLottieSrc} autoplay dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(dotLottieRefCallback).toHaveBeenCalledTimes(1);
    });

    const dotLottie = dotLottieRefCallback.mock.calls[0]?.[0];

    dotLottie?.addEventListener('load', onLoad);

    await vi.waitFor(() => {
      expect(onLoad).toHaveBeenCalledTimes(1);
    });

    const load = vi.spyOn(dotLottie, 'load');

    rerender(<Component src={lottieSrc} autoplay loop speed={2} dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(load).toHaveBeenCalledTimes(1);
    });

    expect(load).toHaveBeenCalledWith({
      src: lottieSrc,
      loop: true,
      autoplay: true,
      speed: 2,
    });
  });

  test('calls dotLottie.setLayout when layout prop changes', async () => {
    const onLoad = vi.fn();
    const dotLottieRefCallback = vi.fn();

    const { rerender } = render(<Component src={dotLottieSrc} autoplay dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(dotLottieRefCallback).toHaveBeenCalledTimes(1);
    });

    const dotLottie = dotLottieRefCallback.mock.calls[0]?.[0];

    dotLottie?.addEventListener('load', onLoad);

    await vi.waitFor(() => {
      expect(onLoad).toHaveBeenCalledTimes(1);
    });

    const setLayout = vi.spyOn(dotLottie, 'setLayout');

    rerender(
      <Component
        src={dotLottieSrc}
        autoplay
        layout={{ align: [0.5, 0.5], fit: 'contain' }}
        dotLottieRefCallback={dotLottieRefCallback}
      />,
    );

    await vi.waitFor(() => {
      expect(setLayout).toHaveBeenCalledTimes(1);
    });

    expect(setLayout).toHaveBeenCalledWith({ align: [0.5, 0.5], fit: 'contain' });

    rerender(<Component src={dotLottieSrc} autoplay dotLottieRefCallback={dotLottieRefCallback} />);

    await vi.waitFor(() => {
      expect(setLayout).toHaveBeenCalledTimes(2);
    });

    expect(setLayout).toHaveBeenCalledWith({ align: [0.5, 0.5], fit: 'contain' });
  });
});
