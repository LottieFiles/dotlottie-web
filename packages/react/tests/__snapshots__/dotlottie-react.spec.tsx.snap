// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`'DotLottieReact' > basic 1`] = `
<div>
  <div
    style="width: 100%; height: 100%; line-height: 0;"
  >
    <canvas
      style="width: 100%; height: 100%;"
    />
  </div>
</div>
`;

exports[`'DotLottieWebGLReact' > basic 1`] = `
<div>
  <div
    style="width: 100%; height: 100%; line-height: 0;"
  >
    <canvas
      style="width: 100%; height: 100%;"
    />
  </div>
</div>
`;

exports[`'DotLottieWebGPUReact' > basic 1`] = `
<div>
  <div
    style="width: 100%; height: 100%; line-height: 0;"
  >
    <canvas
      style="width: 100%; height: 100%;"
    />
  </div>
</div>
`;

exports[`'DotLottieWorkerReact' > basic 1`] = `
<div>
  <div
    style="width: 100%; height: 100%; line-height: 0;"
  >
    <canvas
      style="width: 100%; height: 100%;"
    />
  </div>
</div>
`;
